spark-3.4/clickhouse-spark/src/main/scala/xenon/clickhouse/func/MultiArgsHash.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xenon.clickhouse.func
+
+import org.apache.spark.sql.connector.catalog.functions.{BoundFunction, ScalarFunction, UnboundFunction}
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+abstract class MultiArgsHash extends UnboundFunction with ClickhouseEquivFunction {
+  trait Base extends ScalarFunction[Long] {
+    // must not be private object, nor do it successors, because spark would compile them
+    override def canonicalName: String = s"clickhouse.$name"
+    override def resultType: DataType = LongType
+    override def toString: String = name
+  }
+
+  object Arg1 extends Base {
+    override def name: String = s"${funcName}_1"
+    override def inputTypes: Array[DataType] = Array.fill(1)(StringType)
+    def invoke(value: UTF8String): Long = invokeBase(value)
+  }
+
+  object Arg2 extends Base {
+    override def name: String = s"${funcName}_2"
+    override def inputTypes: Array[DataType] = Array.fill(2)(StringType)
+    def invoke(v1: UTF8String, v2: UTF8String): Long = Seq(v1, v2).map(invokeBase).reduce(combineHashes)
```

Comment:
instead of implementing `def invoke`, there is another option to implement `default R produceResult(InternalRow input)` instead. this approach may address your limitation 4

Reply:
To my understanding, using `produceResult(InternalRow input)` still needs Spark to convert the input to correct type (e.g. `DateType` -> `int`, and we call `getInt` from `InternalRow`), and if this conversion is missing somehow (likely because error says corresponds `invoke` is not found), getting data from `InternalRow` will still be incorrect. On the other hand, if `produceResult` has to handle multiple input types inside, it would still be tricky without knowing them in advance, because we need to call the correct `get_` function on `InternalRow`.

I believe one solution is to make `requiredDistribution ` handles cast expression on the Spark side, and add a type casting in spark clickhouse connector when calling UDFs. 

Please correct me if I misunderstood the procedure.

Reply:
Oops, sorry, I mean the 5 arguments limitation

> Hash functions support for maximum of 5 arguments, limited by Spark's v2 function does not support variable length of inputs

Reply:
Is this possible? I thought `UnboundFunction` needs to bind to a specific data type, then do type casting when input type is not `inputType` attribute of its `BoundFunction`(where var length input is not supported), then actually calling it. Though it will definitely reduce 5 invoke implementation into 1.

Reply:
Pushed a commit supporting varargs, changing `invoke` to `produceResult`. Please take a look when you have time!